api/routers/zip.py

```diff
@@ -0,0 +1,257 @@
+SESSION_TTL_SECONDS = 60 * 30
+
+_zip_sessions: Dict[str, Dict[str, object]] = {}
+_session_lock = Lock()
+
+
+class ZipIngestRequest(BaseModel):
+    zip_id: str = Field(..., description="UUID returned by /upload_zip")
+    selected_files: List[str] = Field(..., min_length=1)
+    dataset_name: str | None = None
+
+
+def _is_supported(filename: str) -> bool:
+    lower = filename.lower()
+    return any(lower.endswith(suffix) for suffix in SUPPORTED_SUFFIXES)
+
+
+def _sanitize_dataset_name(name: str | None) -> str:
+    if not name:
+        return ""
+    sanitized = "".join(ch if ch.isalnum() or ch == "_" else "_" for ch in name)
+    return sanitized.strip("_")
+
+
+async def _write_upload_to_disk(upload: UploadFile, path: Path) -> int:
+    written = 0
+    with open(path, "wb") as buffer:
+        while True:
+            chunk = await upload.read(UPLOAD_CHUNK_SIZE)
+            if not chunk:
+                break
+            buffer.write(chunk)
+            written += len(chunk)
+    return written
+
+
+def _extract_zip(archive: Path, extract_dir: Path) -> Dict[str, str]:
+    files: Dict[str, str] = {}
+    base = extract_dir.resolve()
+
+    with zipfile.ZipFile(archive) as zf:
+        for member in zf.infolist():
+            if member.is_dir():
+                continue
+
+            relative = PurePosixPath(member.filename)
+            if relative.is_absolute() or ".." in relative.parts:
+                raise HTTPException(
+                    status_code=400, detail="ZIP file contains unsafe file paths"
+                )
+
+            dest_path = (extract_dir / Path(relative)).resolve()
+            if not str(dest_path).startswith(str(base)):
+                raise HTTPException(
+                    status_code=400, detail="ZIP file contains unsafe file paths"
+                )
+
+            dest_path.parent.mkdir(parents=True, exist_ok=True)
+            with zf.open(member) as src, open(dest_path, "wb") as dst:
+                shutil.copyfileobj(src, dst, UPLOAD_CHUNK_SIZE)
+
+            files[relative.as_posix()] = str(dest_path)
+
+    return files
+
+
+def _prune_sessions() -> None:
+    cutoff = time.time() - SESSION_TTL_SECONDS
+    expired: List[str] = []
+    for key, session in list(_zip_sessions.items()):
+        created = session.get("created_at")
+        if isinstance(created, (int, float)) and created < cutoff:
+            expired.append(key)
+
+    for key in expired:
+        session = _zip_sessions.pop(key, None)
+        if not session:
+            continue
+        base_dir = Path(session.get("base_dir", ""))
+        if base_dir.exists():
+            shutil.rmtree(base_dir, ignore_errors=True)
+
+
+def _store_session(zip_id: str, payload: Dict[str, object]) -> None:
+    with _session_lock:
+        _prune_sessions()
+        _zip_sessions[zip_id] = payload
+
+
+def _get_session(zip_id: str) -> Dict[str, object]:
+    with _session_lock:
+        _prune_sessions()
```

Comment:
why do we need _prune_sessions during this read? it makes sense to clean up old sessions when adding a new one in `_store_session` but here we could probably remove unless there was a good reason.